### PR TITLE
[FLOC-1674] Switch to conch

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1674

### CODE REVIEW

File: flocker/provision/_effect.py [hunk 1]
```diff
@@ -17,6 +17,9 @@ class SequenceFailed(Exception, object):
     :ivar error: The error result of the last run effect.
     """
 
+    def __str__(self):
+        return repr(self)
```
Comment (wallrj) [inline flocker/provision/_effect.py]: Isn't this Python's default behaviour?

```
In [1]: class Foo(object):
   ...:     def __repr__(self):
   ...:         return 'I am Foo'
   ...:     

In [2]: Foo()
Out[2]: I am Foo

In [3]: str(Foo())
Out[3]: 'I am Foo'

```

File: flocker/provision/_ssh/_conch.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+@attributes([
+    "deferred",
+    "context",
+])
+class CommandProtocol(LineOnlyReceiver, object):
+    """
+    Protocol that logs the lines of a remote command.
+
+    :ivar Deferred deferred: Deferred to fire when the command finishes
+        If the command finished succesfuly, will fire with ``None``.
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: Typo "succesfuly"

File: flocker/provision/_ssh/_conch.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        Otherwise, errbacks with the reason.
+    :ivar bytes username: For logging.
+    :ivar bytes address: For logging.
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: Looks like these attributes have now been replaced with `context`.

File: flocker/provision/_ssh/_conch.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+def perform_comment(dispatcher, intent):
+    """
+    See :py:class:`Comment`.
+    """
+
+
+def get_ssh_dispatcher(connection, context):
+    """
+    :ivar bytes username: For logging.
+    :ivar bytes address: For logging.
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: Parameter docs need updating.

File: flocker/provision/test/test_ssh_conch.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    def setUp(self):
+        self.sshd_config = FilePath(self.mktemp())
+        self.server = create_ssh_server(self.sshd_config)
+        self.addCleanup(self.server.restore)
+
+        self.agent = create_ssh_agent(self.server.key_path)
+        self.addCleanup(self.agent.restore)
+
+    def test_stuff(self):
```
Comment (wallrj) [inline flocker/provision/test/test_ssh_conch.py]: Maybe remove this or rename and document it.

File: flocker/provision/test/test_ssh_conch.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        )
+
+        d = perform(
+            make_dispatcher(reactor),
+            command,
+        )
+        return d
+
+    def test_put(self):
+
```
Comment (wallrj) [inline flocker/provision/test/test_ssh_conch.py]: Docstring.

**Comment by wallrj** [comment]
I got the following error message when I tried `run-acceptance-tests` without first loading my key...

```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ time ./admin/run-acceptance-tests --keep --distribution=centos-7 --provider=aws --config-file=aws.config --branch=0.4.0 --flocker-version=0.4.0 --variant=docker-head |& tee switch-to-conch-FLOC-1674.log
main function encountered error
Traceback (most recent call last):
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/effect/_base.py", line 151, in perform
    trampoline(_perform, effect)
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/effect/_continuation.py", line 61, in trampoline
    return
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/effect/_base.py", line 134, in _run_callbacks
    bouncer.bounce(_run_callbacks, chain, result)
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/effect/_base.py", line 83, in guard
    return (True, sys.exc_info())
--- <exception caught here> ---
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/effect/_base.py", line 81, in guard
    return (False, f(*args, **kwargs))
  File "/home/richard/projects/HybridLogic/flocker/flocker/provision/_effect.py", line 70, in fail
    result[2])
  File "<string>", line 2, in reraise
    
effect._intents.FirstError: (index=0) SequenceFailed: <SequenceFailed(results=[], exc_info=(<class 'twisted.conch.ssh.keys.EncryptedKeyError'>, EncryptedKeyError('Passphrase must be provided for an encrypted key',), None))>
Creating node 0: acceptance-test-richardw-0
Creating node 1: acceptance-test-richardw-1
--keep specified, not destroying nodes.

```

Which is fine but by that time it's already started the two nodes.
It'd be even nicer if it could check that the configured key was loaded up front....but not not necessary here.

**Comment by exarkun** [comment]
> Which is fine but by that time it's already started the two nodes.
> It'd be even nicer if it could check that the configured key was loaded up front....but not not necessary here.

Actually... Considering how much time we spend waiting on builds, I'd lean towards saying simple changes like this that significantly cut down on waiting time are worth doing now.

**Comment by wallrj** [comment]
Hmm, I get the same error even after loading my AWS  key 

```
effect._intents.FirstError: (index=0) SequenceFailed: <SequenceFailed(results=[], exc_info=(<class 'twisted.conch.ssh.keys.EncryptedKeyError'>, EncryptedKeyError('Passphrase must be provided for an encrypted key',), None))>
Creating node 0: acceptance-test-richardw-0
Creating node 1: acceptance-test-richardw-1
--keep specified, not destroying nodes.

real	1m51.944s
user	0m39.155s
sys	0m0.613s
```


```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ ssh-add  -l
2048 e0:b6:32:e4:25:46:c8:c9:ae:8c:63:4f:87:3f:a5:b7 /home/richard/.ssh/richardw-testing.pem (RSA)
...
```

```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ grep keyname aws.config 
  keyname: "richardw-testing"

```
 
...and I am able to connect using openssh...

```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ ssh centos@52.11.51.215
[centos@ip-172-31-39-219 ~]$ 
```

**Comment by wallrj** [comment]
Tried running using centos-7 + vagrant but got `distribution not found`....is that because the centos-7 tutorial box hasn't landed yet?

**Comment by wallrj** [comment]
Tried running with Fedora-20 + vagrant but got and error...because I'd supplied a variant (perhaps that's a separate issue)

```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ time ./admin/run-acceptance-tests --keep --distribution=fedora-20 --provider=vagrant --branch=0.4.0 --flocker-version=0.4.0 --variant=docker-head |& tee switch-to-conch-FLOC-1674.log
main function encountered error
Traceback (most recent call last):
  File "./admin/run-acceptance-tests", line 14, in <module>
    react(main, (sys.argv[1:], BASEPATH, TOPLEVEL))
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/twisted/internet/task.py", line 875, in react
    finished = main(_reactor, *argv)
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/twisted/internet/defer.py", line 1253, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
--- <exception caught here> ---
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/twisted/internet/defer.py", line 1107, in _inlineCallbacks
    result = g.send(result)
  File "/home/richard/projects/HybridLogic/flocker/admin/acceptance.py", line 398, in main
    options.parseOptions(args)
  File "/home/richard/.virtualenvs/1674/lib/python2.7/site-packages/twisted/python/usage.py", line 277, in parseOptions
    self.postOptions()
  File "/home/richard/projects/HybridLogic/flocker/admin/acceptance.py", line 370, in postOptions
    variants=self['variants'],
  File "<characteristic generated init c1cd63332a78e95f170921236217c9e5772551eb>", line 18, in characteristic_init
    self.__original_init__(*args, **kw)
  File "/home/richard/projects/HybridLogic/flocker/admin/acceptance.py", line 150, in __init__
    % (', '.join(self.variants),))
exceptions.TypeError: sequence item 0: expected string, ValueConstant found

real	0m0.930s
user	0m0.887s
sys	0m0.046s

```

**Comment by wallrj** [comment]
Sorry @tomprince, I get an error with vagrant provider too.

```
(1674)(switch-to-conch-FLOC-1674 ? =)[~/.../HybridLogic/flocker]$ time ./admin/run-acceptance-tests --keep --distribution=fedora-20 --provider=vagrant --branch=0.4.0 --flocker-version=0.4.0 |& tee switch-to-conch-FLOC-1674.log
==> node2: Destroying VM and associated drives...
==> node1: Destroying VM and associated drives...
Bringing machine 'node1' up with 'virtualbox' provider...
Bringing machine 'node2' up with 'virtualbox' provider...
==> node1: Importing base box 'clusterhq/flocker-tutorial'...
==> node1: Matching MAC address for NAT networking...
==> node1: Checking if box 'clusterhq/flocker-tutorial' is up to date...
==> node1: Setting the name of the VM: fedora-20_node1_1429530941321_85506
==> node1: Fixed port collision for 22 => 2222. Now on port 2200.
==> node1: Clearing any previously set network interfaces...
==> node1: Preparing network interfaces based on configuration...
    node1: Adapter 1: nat
    node1: Adapter 2: hostonly
==> node1: Forwarding ports...
    node1: 22 => 2200 (adapter 1)
==> node1: Booting VM...
==> node1: Waiting for machine to boot. This may take a few minutes...
    node1: SSH address: 127.0.0.1:2200
    node1: SSH username: vagrant
    node1: SSH auth method: private key
    node1: Warning: Connection timeout. Retrying...
    node1: 
    node1: Vagrant insecure key detected. Vagrant will automatically replace
    node1: this with a newly generated keypair for better security.
    node1: 
    node1: Inserting generated public key within guest...
    node1: Removing insecure key from the guest if its present...
    node1: Key inserted! Disconnecting and reconnecting using new SSH key...
==> node1: Machine booted and ready!
==> node1: Checking for guest additions in VM...
==> node1: Setting hostname...
==> node1: Configuring and enabling network interfaces...
==> node1: Mounting shared folders...
    node1: /vagrant => /home/richard/projects/HybridLogic/flocker/admin/vagrant-acceptance-targets/fedora-20
==> node2: Importing base box 'clusterhq/flocker-tutorial'...
==> node2: Matching MAC address for NAT networking...
==> node2: Checking if box 'clusterhq/flocker-tutorial' is up to date...
==> node2: Setting the name of the VM: fedora-20_node2_1429531103858_18293
==> node2: Clearing any previously set network interfaces...
==> node2: Preparing network interfaces based on configuration...
    node2: Adapter 1: nat
    node2: Adapter 2: hostonly
==> node2: Forwarding ports...
    node2: 22 => 2222 (adapter 1)
==> node2: Booting VM...
==> node2: Waiting for machine to boot. This may take a few minutes...
    node2: SSH address: 127.0.0.1:2222
    node2: SSH username: vagrant
    node2: SSH auth method: private key
    node2: Warning: Connection timeout. Retrying...
    node2: 
    node2: Vagrant insecure key detected. Vagrant will automatically replace
    node2: this with a newly generated keypair for better security.
    node2: 
    node2: Inserting generated public key within guest...
    node2: Removing insecure key from the guest if its present...
    node2: Key inserted! Disconnecting and reconnecting using new SSH key...
==> node2: Machine booted and ready!
==> node2: Checking for guest additions in VM...
==> node2: Setting hostname...
==> node2: Configuring and enabling network interfaces...
==> node2: Mounting shared folders...
    node2: /vagrant => /home/richard/projects/HybridLogic/flocker/admin/vagrant-acceptance-targets/fedora-20
# Host 172.16.255.240 found: line 26 type ECDSA
main function encountered error
Traceback (most recent call last):
Failure: twisted.conch.ssh.keys.EncryptedKeyError: Passphrase must be provided for an encrypted key
--keep specified, not destroying nodes.

real	5m59.372s
user	0m11.606s
sys	0m6.770s
```

Perhaps I've got some dodgy lines in my known_hosts file?

**Comment by wallrj** [comment]
Oh, I forgot to push the jira button.

The code changes all look good to me, but I think I'll reject this for now since I haven't been able to run the tool.

I'll take another look when the comments and issues above have been addressed.

**Comment by tomprince** [comment]
>> It'd be even nicer if it could check that the configured key was loaded up front....but not not necessary here.

> Actually... Considering how much time we spend waiting on builds, I'd lean towards saying simple changes like this that significantly cut down on waiting time are worth doing now.

That seems like a sensible change, but it seems worth doing in a separate ticket, since it seems somewhat orthogonal to the changes here.

**Comment by tomprince** [comment]
For the vagrant box, you'll need the insecure-key in you agent.

File: admin/acceptance.py [hunk 7]
```diff
@@ -370,6 +375,18 @@ def postOptions(self):
             )
 
 
+def eliot_output(message):
+    if message['message_type'] == "flocker.provision.ssh:run":
```
Comment (wallrj) [inline admin/acceptance.py]: These would be neater in a dict `{message_type: message_format}`

File: admin/acceptance.py [hunk 8]
```diff
@@ -370,6 +375,18 @@ def postOptions(self):
             )
 
 
+def eliot_output(message):
```
Comment (wallrj) [inline admin/acceptance.py]: Add a short docstring.

File: flocker/provision/_aws.py [hunk 9]
```diff
@@ -117,6 +119,10 @@ def create_arguments(disk_size):
                          "DeleteOnTermination": True,
                          "VolumeType": "gp2"}}
             ],
+            "ex_userdata": dedent("""\
+                #!/bin/sh
+                sed -i '/Defaults *requiretty/d' /etc/sudoers
+                """)
```
Comment (wallrj) [inline flocker/provision/_aws.py]: I remember seeing a discussion about this on zulip. Add an explanatory comment.

File: flocker/provision/_ssh/_conch.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+import os
+
+from flocker.testtools import loop_until
+
+from ._model import Run, Sudo, Put, Comment, RunRemotely
+
+from .._effect import dispatcher as base_dispatcher
+
+
+logger = Logger()
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: The pattern elsewhere is to make this a private module variable:
```
(master ? =)[~/.../HybridLogic/flocker]$ ack '_logger = Logger()'
flocker/node/agents/blockdevice.py
29-# Eliot is transitioning away from the "Logger instances all over the place"
30-# approach.  And it's hard to put Logger instances on PRecord subclasses which
31-# we have a lot of.  So just use this global logger for now.
32:_logger = Logger()

flocker/node/_deploy.py
35:_logger = Logger()

flocker/restapi/_infrastructure.py
30:_logger = Logger()

```

File: flocker/provision/_ssh/_conch.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+@sync_performer
+def perform_comment(dispatcher, intent):
+    """
+    See :py:class:`Comment`.
+    """
+
+
+def get_ssh_dispatcher(connection, context):
+    """
+    :ivar Message context: The eliot message context to log.
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: `:param ...:`

File: flocker/provision/_ssh/_conch.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+from effect import (
+    sync_performer, TypeDispatcher, ComposedDispatcher, Effect,
+    )
+from effect.twisted import (
+    make_twisted_dispatcher,
+)
+from effect.twisted import (
+    perform, deferred_performer)
+from twisted.conch.endpoints import (
+    SSHCommandClientEndpoint, _NewConnectionHelper, _ReadFile, ConsoleUI)
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: Link to Twisted tickets to make these public...as just discussed.

File: flocker/provision/_ssh/_conch.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    Get a :class:`twisted.conch.endpoints._ISSHConnectionCreator` to connect to
+    the given remote.
+
+    :param bytes address: The address of the remote host to connect to.
+    :param bytes username: The user to connect as.
+    :param int port: The port of the ssh server to connect to.
+
+    :return _ISSHConnectionCreator:
+    """
+    key_path = FilePath(os.path.expanduser('~/.ssh/id_rsa'))
```
Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: I'll try removing this code and re-running, as discussed.

Comment (wallrj) [inline flocker/provision/_ssh/_conch.py]: Yep, removing this allows the script to progress.

File: flocker/provision/_ssh/_conch.py (hunk 12 above)
Comment (exarkun) [inline flocker/provision/_ssh/_conch.py]: I'm not sure how likely it is that Twisted will make these public.

File: flocker/provision/_ssh/_conch.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    except KeyError:
+        agentEndpoint = None
+
+    return _NewConnectionHelper(
+        reactor, address, port, None, username,
+        keys=None,
+        password=None,
+        agentEndpoint=agentEndpoint,
+        knownHosts=KnownHostsFile.fromPath(FilePath("/dev/null")),
+        ui=ConsoleUI(lambda: _ReadFile(b"yes")))
```
Comment (exarkun) [inline flocker/provision/_ssh/_conch.py]: Or

```
class Yes(object):
    def prompt(self, text):
        return succeed(True)

    def warn(self, text):
        pass
```

and then you don't need `_ReadFile`.

File: flocker/provision/_ssh/_conch.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+    :return _ISSHConnectionCreator:
+    """
+    try:
+        agentEndpoint = UNIXClientEndpoint(
+            reactor, os.environ["SSH_AUTH_SOCK"])
+    except KeyError:
+        agentEndpoint = None
+
+    return _NewConnectionHelper(
```
Comment (exarkun) [inline flocker/provision/_ssh/_conch.py]: Or do what the conch shared endpoint example does and use `newConnection` as a factory for the connection to be used when you want to use `existingConnection`.

Comment (exarkun) [inline flocker/provision/_ssh/_conch.py]: (And then you don't need `_NewConnectionHelper`.)

Comment (tomprince) [inline flocker/provision/_ssh/_conch.py]: Running a `cat` process just because this isn't yet public bothers me.

Comment (exarkun) [inline flocker/provision/_ssh/_conch.py]: Assuming that "yet" actually works in that sentence bothers me. :smile: 

The extra process is entirely harmless.  Using a private API from another project is potentially harmful.

**Comment by wallrj** [comment]
Ok. Well having removed the code that loads id_rsa, the script once again runs for me.

I tried to compare the times but I made the mistake of running specifying the package 0.4.0 which isn't compatible with the `_compose` API version of `flocker-deploy` in my working branch.

But the new logging is much nicer and the parallel installation of ZFS on cloud node must give a significant speed up.


So please address or answer the comments above and then merge.

File: flocker/provision/_ssh/_conch.py (hunk 15 above)
Comment (tomprince) [inline flocker/provision/_ssh/_conch.py]: I spent more time than is warranted trying to get the tests to pass with a clean reactor. I think I may have narrowed it down to `flocker.testtools.ssh.create_ssh_server(...).restore()` not waiting for any open connections to close.

**Comment by tomprince** [comment]
The homebrew test fails because $PATH is set only for interactive logins, which fabric does, but our method of using conch doesn't. I'm not sure how to address this.

**Pull request merged**